File: scripts/sentry-pull.py
#!/usr/bin/env python3
"""Pull Sentry errors for SkaiScraper production."""
import json
import urllib.request
import urllib.error
import sys
import os

TOKEN = "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"
ORG = "clearskai-technologies"
PROJECT = "javascript-nextjs"

def fetch_issues(period="24h", query="is:unresolved"):
    url = f"https://us.sentry.io/api/0/projects/{ORG}/{PROJECT}/issues/?query={query}&statsPeriod={period}&sort=freq"
    req = urllib.request.Request(url, headers={"Authorization": f"Bearer {TOKEN}"})
    try:
        with urllib.request.urlopen(req) as resp:
            return json.loads(resp.read())
    except urllib.error.HTTPError as e:
        body = e.read().decode()
        print(f"Sentry API error {e.code}: {body[:300]}")
        return None

def fetch_issue_events(issue_id, limit=3):
    url = f"https://us.sentry.io/api/0/issues/{issue_id}/events/?full=true"
    req = urllib.request.Request(url, headers={"Authorization": f"Bearer {TOKEN}"})
    try:
        with urllib.request.urlopen(req) as resp:
            events = json.loads(resp.read())
            return events[:limit]
    except:
        return []

print("=" * 60)
print("  SENTRY ERROR REPORT — SkaiScraper Production")
print("=" * 60)

# Try last 24h first, then expand to 14d if empty
for period, label in [("24h", "Last 24 Hours"), ("14d", "Last 14 Days")]:
    print(f"\n--- {label} (Unresolved) ---")
    issues = fetch_issues(period=period)
    if issues is None:
        sys.exit(1)
    if not issues:
        print(f"  No unresolved issues in {label}!")
        if period == "24h":
            continue
        else:
            break
    
    print(f"  Found {len(issues)} unresolved issues:\n")
    for i, issue in enumerate(issues[:25], 1):
        title = issue.get("title", "Unknown")
        culprit = issue.get("culprit", "")
        count = issue.get("count", "?")
        level = issue.get("level", "?")
        first = issue.get("firstSeen", "")[:19]
        last = issue.get("lastSeen", "")[:19]
        link = issue.get("permalink", "")
        
        print(f"  {i:2}. [{level.upper()}] {title}")
        print(f"      Culprit: {culprit}")
        print(f"      Count: {count} | First: {first} | Last: {last}")
        
        # Get stack trace for top errors
        if i <= 10:
            events = fetch_issue_events(issue.get("id"))
            for ev in events[:1]:
                entries = ev.get("entries", [])
                for entry in entries:
                    if entry.get("type") == "exception":
                        values = entry.get("data", {}).get("values", [])
                        for exc in values[:1]:
                            exc_type = exc.get("type", "")
                            exc_value = exc.get("value", "")[:200]
                            print(f"      Exception: {exc_type}: {exc_value}")
                            frames = exc.get("stacktrace", {}).get("frames", [])
                            for frame in frames[-3:]:
                                fn = frame.get("function", "?")
                                filename = frame.get("filename", "?")
                                lineno = frame.get("lineNo", "?")
                                print(f"        → {filename}:{lineno} in {fn}")
        print(f"      Link: {link}")
        print()
    break  # Don't do 14d if 24h had results

# Also fetch errors specifically
print("\n--- Error-Level Issues (14d) ---")
error_issues = fetch_issues(period="14d", query="is:unresolved level:error")
if error_issues:
    print(f"  {len(error_issues)} error-level issues")
    for i, issue in enumerate(error_issues[:15], 1):
        title = issue.get("title", "Unknown")
        count = issue.get("count", "?")
        last = issue.get("lastSeen", "")[:19]
        print(f"  {i:2}. {title} (count: {count}, last: {last})")
else:
    print("  No error-level issues!")

print("\n" + "=" * 60)
print("  REPORT COMPLETE")
print("=" * 60)

File: scripts/sentry-pull.mjs
// Pull Sentry errors for SkaiScraper production
const https = require("https");

const TOKEN =
  "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";
const ORG = "clearskai-technologies";
const PROJECT = "javascript-nextjs";

function fetchJSON(url) {
  return new Promise((resolve, reject) => {
    const req = https.get(url, { headers: { Authorization: `Bearer ${TOKEN}` } }, (res) => {
      let data = "";
      res.on("data", (chunk) => (data += chunk));
      res.on("end", () => {
        try {
          resolve({ status: res.statusCode, body: JSON.parse(data) });
        } catch (e) {
          resolve({ status: res.statusCode, body: data.slice(0, 500) });
        }
      });
    });
    req.on("error", reject);
  });
}

async function main() {
  console.log("=".repeat(60));
  console.log("  SENTRY ERROR REPORT — SkaiScraper Production");
  console.log("=".repeat(60));

  // Try 24h first
  for (const [period, label] of [
    ["24h", "Last 24 Hours"],
    ["14d", "Last 14 Days"],
  ]) {
    console.log(`\n--- ${label} (Unresolved) ---`);
    const url = `https://us.sentry.io/api/0/projects/${ORG}/${PROJECT}/issues/?query=is%3Aunresolved&statsPeriod=${period}&sort=freq`;
    const { status, body } = await fetchJSON(url);

    if (status !== 200) {
      console.log(
        `  API Error ${status}:`,
        typeof body === "string" ? body : JSON.stringify(body).slice(0, 300)
      );
      continue;
    }

    if (!Array.isArray(body) || body.length === 0) {
      console.log(`  No unresolved issues in ${label}!`);
      if (period === "24h") continue;
      break;
    }

    console.log(`  Found ${body.length} unresolved issues:\n`);
    for (let i = 0; i < Math.min(body.length, 25); i++) {
      const issue = body[i];
      console.log(`  ${i + 1}. [${(issue.level || "?").toUpperCase()}] ${issue.title}`);
      console.log(`     Culprit: ${issue.culprit || "N/A"}`);
      console.log(
        `     Count: ${issue.count} | First: ${(issue.firstSeen || "").slice(0, 19)} | Last: ${(issue.lastSeen || "").slice(0, 19)}`
      );

      // Get stack trace for top 10
      if (i < 10) {
        try {
          const evUrl = `https://us.sentry.io/api/0/issues/${issue.id}/events/?full=true`;
          const evRes = await fetchJSON(evUrl);
          if (evRes.status === 200 && Array.isArray(evRes.body) && evRes.body[0]) {
            const ev = evRes.body[0];
            for (const entry of ev.entries || []) {
              if (entry.type === "exception") {
                for (const exc of (entry.data?.values || []).slice(0, 1)) {
                  console.log(`     Exception: ${exc.type}: ${(exc.value || "").slice(0, 200)}`);
                  const frames = (exc.stacktrace?.frames || []).slice(-3);
                  for (const f of frames) {
                    console.log(`       → ${f.filename}:${f.lineNo} in ${f.function}`);
                  }
                }
              }
            }
          }
        } catch (e) {
          /* skip event details */
        }
      }
      console.log(`     Link: ${issue.permalink || ""}`);
      console.log();
    }
    break; // Don't expand if we found results
  }

  // Also get error-level specifically
  console.log("\n--- Error-Level Issues (14d) ---");
  const errUrl = `https://us.sentry.io/api/0/projects/${ORG}/${PROJECT}/issues/?query=is%3Aunresolved%20level%3Aerror&statsPeriod=14d&sort=freq`;
  const errRes = await fetchJSON(errUrl);
  if (errRes.status === 200 && Array.isArray(errRes.body)) {
    if (errRes.body.length === 0) {
      console.log("  No error-level issues!");
    } else {
      console.log(`  ${errRes.body.length} error-level issues:`);
      for (let i = 0; i < Math.min(errRes.body.length, 15); i++) {
        const issue = errRes.body[i];
        console.log(
          `  ${i + 1}. ${issue.title} (count: ${issue.count}, last: ${(issue.lastSeen || "").slice(0, 19)})`
        );
      }
    }
  }

  console.log("\n" + "=".repeat(60));
  console.log("  REPORT COMPLETE");
  console.log("=".repeat(60));
}

main().catch((e) => console.error("Fatal:", e));
